In [46]:
import os
import json

import requests
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [ ]:
OPENAI_API_KEY = os.environ["OPENAI_API_KEY"]

In [2]:
cgfp = pd.read_csv("~/Box/dsi-core/11th-hour/good-food-purchasing/CONFIDENTIAL_GFPP Product Attribute List_8.26.25.csv", dtype=str)

In [3]:
cgfp[["Product GTIN or UPC", "Vendor", "Brand Name", "Product Type", "Food Product Category", "Product Name", "Level of Processing"]]

Product GTIN or UPC           Vendor     Brand Name  \
0                     NaN            Sysco            NaN   
1                     NaN  Produce Express            NaN   
2                     NaN  Produce Express            NaN   
3                     NaN  Produce Express            NaN   
4                     NaN  Produce Express            NaN   
...                   ...              ...            ...   
90482                 NaN   Thurston Foods  UPSTATE FARMS   
90483                 NaN   Thurston Foods  UPSTATE FARMS   
90484                 NaN   Thurston Foods       DANIMALS   
90485                 NaN   Thurston Foods  UPSTATE FARMS   
90486                 NaN   Thurston Foods  UPSTATE FARMS   

                                 Product Type Food Product Category  \
0                 FRUIT CUP PEACH COMDTY A422                 Fruit   
1               ASPARAGUS 11# PENCIL (MEXICO)            Vegetables   
2               CELERY 24'S (SANTA MARIA, CA)            Vegetables   
3                   ENGLISH CUCUMBER 12 COUNT            Vegetables   
4                        PERSIAN CUCUMBER 22#            Vegetables   
...                                       ...                   ...   
90482          YOGURT VANILLA NF GF (3050000)                Yogurt   
90483           YOGURT BLUEBERRY NF (3050400)                Yogurt   
90484  YOGURT VANILLA DANIMAL NF GF (3051300)                Yogurt   
90485   YOGURT CHERRY VANILLA NF GF (3051900)                Yogurt   
90486            YOGURT PEACH NF GF (3052300)                Yogurt   

                      Product Name        Level of Processing  
0      fruit cup, peach, commodity            Ultra-Processed  
1                        asparagus  Whole/Minimally Processed  
2                           celery  Whole/Minimally Processed  
3                cucumber, english  Whole/Minimally Processed  
4                cucumber, persian  Whole/Minimally Processed  
...                            ...                        ...  
90482     yogurt, flavored, nonfat            Ultra-Processed  
90483     yogurt, flavored, nonfat            Ultra-Processed  
90484     yogurt, flavored, nonfat            Ultra-Processed  
90485     yogurt, flavored, nonfat            Ultra-Processed  
90486     yogurt, flavored, nonfat            Ultra-Processed  

[90487 rows x 7 columns]

In [4]:
cgfp_with_upc = cgfp[cgfp["Product GTIN or UPC"].notna() & ~cgfp["Product GTIN or UPC"].isin(["#REF!"])].copy()
cgfp_with_upc["Product GTIN or UPC"] = cgfp_with_upc["Product GTIN or UPC"].apply(lambda x: f"{int(float(x.replace(' ', ''))):014d}")

In [37]:
cgfp_mini = cgfp_with_upc[["Product GTIN or UPC", "Vendor", "Brand Name", "Product Type", "Food Product Category", "Product Name", "Level of Processing"]]
cgfp_mini

Product GTIN or UPC                    Vendor          Brand Name  \
84         10731149515854              Sysco Hawaii                 NaN   
85         00044978102502              Sysco Hawaii                 NaN   
86         10027000383091              Sysco Hawaii                 NaN   
87         10876681007822              Sysco Hawaii                 NaN   
88         10889356022952              Sysco Hawaii                 NaN   
...                   ...                       ...                 ...   
89392      10078800113000  Performance Food Service       UPSTATE FARMS   
89393      10070470311000  Performance Food Service             YOPLAIT   
89394      10070470166000  Performance Food Service  YOPLAIT PARFAITPRO   
89395      10078800113000  Performance Food Service       UPSTATE FARMS   
89396      20036632005000  Performance Food Service  DANNON LIGHT & FIT   

                                  Product Type Food Product Category  \
84          FISH SALMON CHUM FILLET SK/ON 2-3#           Fish (Wild)   
85           RICE JASMINE GOLDEN PHOENIX/CAMEL                  Rice   
86            KETCHUP TOMATO POUCH PK 114Z NTF   Condiments & Snacks   
87        FLATBREAD NAAN TEARDROP ORIG (6-8CT)        Grain Products   
88     OLIVES GREEN QUEEN STFD PIMENTO 100/120   Condiments & Snacks   
...                                        ...                   ...   
89392                   YOGURT STRWBRY BLND NF                Yogurt   
89393              YOGURT TRIX TRIPLE CHRY TFF                Yogurt   
89394           YOGURT VANILLA LF POUCH TFF GF                Yogurt   
89395                        YOGURT VANILLA NF                Yogurt   
89396                   YOGURT VAR PACK CUP NF                Yogurt   

                       Product Name        Level of Processing  
84                     salmon, wild  Whole/Minimally Processed  
85                    rice, jasmine  Whole/Minimally Processed  
86           condiment, ketchup, ss            Ultra-Processed  
87                 bread, flatbread            Ultra-Processed  
88                   olive, stuffed            Ultra-Processed  
...                             ...                        ...  
89392              yogurt, flavored            Ultra-Processed  
89393              yogurt, flavored            Ultra-Processed  
89394              yogurt, flavored            Ultra-Processed  
89395      yogurt, flavored, nonfat            Ultra-Processed  
89396  yogurt, flavored, nonfat, ss            Ultra-Processed  

[3432 rows x 7 columns]

In [6]:
off = pd.read_csv("~/Box/dsi-core/11th-hour/good-food-purchasing/en.openfoodfacts.org.products.csv.gz", sep="\t", dtype=str, usecols=range(60))
off["code"] = off["code"].apply(lambda x: f"{np.iinfo(np.uint64).max if len(x) >= 20 else np.uint64(x):014d}")

In [98]:
len(off)

4014198

In [35]:
off_mini = off[["code", "product_name", "brands", "categories", "ingredients_text", "nova_group"]]
off_mini.iloc[2000000:2000100]

code                                       product_name  \
2000000  03760284500067  Crème glacée Praliné aux éclats d'amendes cara...   
2000001  03760284700009                                   Baguette de pain   
2000002  03760284700016                                 Baguette au seigle   
2000003  03760284700030                              Baguette complète bio   
2000004  03760284700061                               Pavé des copains Bio   
...                 ...                                                ...   
2000095  03760285379532                                                NaN   
2000096  03760285410013            Tuiles de sarrasin aux 2 graines de lin   
2000097  03760285410020   Tuiles de sarrasin à l’algue sauvage de bretagne   
2000098  03760285410037                                 Tuiles de sarrasin   
2000099  03760285410044                                 Tuiles de sarrasin   

        brands     categories ingredients_text nova_group  
2000000    NaN            NaN              NaN        NaN  
2000001    NaN            NaN              NaN        NaN  
2000002    NaN            NaN              NaN        NaN  
2000003    NaN            NaN              NaN        NaN  
2000004    NaN            NaN              NaN        NaN  
...        ...            ...              ...        ...  
2000095    NaN     Bières IPA              NaN        NaN  
2000096    NaN            NaN              NaN        NaN  
2000097    NaN            NaN              NaN        NaN  
2000098    NaN  Tuiles salées              NaN        NaN  
2000099    NaN            NaN              NaN        NaN  

[100 rows x 6 columns]

In [26]:
merged = off_mini.merge(cgfp_mini, left_on="code", right_on="Product GTIN or UPC")

In [51]:
len(merged)

364

In [52]:
is_good_match = []
for i, (_, row) in enumerate(merged.iterrows()):
    try:
        response = requests.post(
            "https://api.openai.com/v1/chat/completions",
            headers={
                "Content-Type": "application/json",
                "Authorization": f"Bearer {OPENAI_API_KEY}",
            },
            json={
                "model": "gpt-4.1",
                "messages": [
                    {"role": "system", "content": """
You are comparing food products from two databases. Return `{"is_good_match": true}`
if the food product described by the first group of data fields is like the same
food product as described by the second group.
""".strip()},
            {"role": "user", "content": f"""
First group:
  * product_name: {row['product_name']}
  * brands: {row['brands']}
  * categories: {row['categories']}
  * ingredients: {row['ingredients_text']}

Second group:
  * Vendor: {row['Vendor']}
  * Brand Name: {row['Brand Name']}
  * Product Type: {row['Product Type']}
  * Food Product Category: {row['Food Product Category']}
  * Product Name: {row['Product Name']}
""".strip()},
                ],
                "response_format": {
                    "type": "json_schema",
                    "json_schema": {
                        "name": "is_good_match",
                        "schema": {
                            "type": "object",
                            "properties": {
                                "is_good_match": {"type": "boolean"},
                            },
                            "required": ["is_good_match"],
                            "additionalProperties": False,
                        },
                    },
                },
            },
        )
        is_good_match.append(json.loads(response.json()["choices"][0]["message"]["content"])["is_good_match"])
    except Exception:
        is_good_match.append(None)
    print(f"{i} {row['product_name']} =?= {row['Brand Name']} ? {is_good_match[-1]}")

0 The Smartest Cookie =?= nan ? False
1 Hershey’s Syrup =?= nan ? False
2 nan =?= nan ? False
3 Vegan plus =?= nan ? False
4 Vegan plus =?= nan ? False
5 Nectar de cerise =?= nan ? False
6 Probiotic =?= nan ? False
7 Whey Protein - Chocolate =?= nan ? False
8 Skinny juice =?= nan ? False
9 No chicken kiyvs =?= nan ? False
10 McCroustillant bang bang - Mc Donalds =?= nan ? False
11 ISO whey zero =?= nan ? False
12 Miel de Forêt =?= nan ? False
13 Curry powder =?= nan ? False
14 Hennesy =?= nan ? False
15 Metaxa Private Reserve =?= nan ? False
16 nan =?= nan ? False
17 Apple Turnover =?= nan ? False
18 nan =?= nan ? False
19 ConfiNoix =?= nan ? True
20 Petit hoppelli noir =?= nan ? False
21 Mamma Mia  American BBQ =?= nan ? False
22 Fit Piggy Snickers Twist =?= nan ? False
23 Fit Piggy Brownie =?= nan ? False
24 Confiture de  chataigne Bio =?= nan ? False
25 RENE BENDEJA NUECES MACADAMIA =?= nan ? False
26 Rosmarin Kartoffeln Maggi =?= nan ? False
27 Organic Oats =?= nan ? False
28 Coop 

In [53]:
merged["is_good_match"] = is_good_match

In [93]:
merged["nova_group"] = merged["nova_group"].map({"1": 1, "2": 2, "3": 3, "4": 4})

In [79]:
merged["Level of Processing"] = merged["Level of Processing"].map({"Whole/Minimally Processed": 1, "Culinary Ingredient": 2, "Moderately Processed": 3, "Ultra-Processed": 4})

In [99]:
len(merged.query("is_good_match"))

286

The first four columns, `product_name`, `brands`, `categories`, `ingredients_list`, come from Open Food Facts.

The last five columns, `Vendor`, `Brand Name`, `Product Type`, `Food Product Category`, `Product Name`, come from CGFP.

They should agree because I'm filtering on ChatGPT's determination that they are the same food product.

In [80]:
merged.query("is_good_match")[["product_name", "brands", "categories", "ingredients_text", "Vendor", "Brand Name", "Product Type", "Food Product Category", "Product Name"]].iloc[60:120]

product_name  \
128             Cheez it grooves white cheddar peg bag   
129                    Cheez it Bacon & Cheddar Cheese   
130                              Cheez It Extra Cheesy   
131                             Cheez it White Cheddar   
132                         Cheez-It Crackers Original   
133                                  Cheez It Original   
134                         Cheez It Hot & Spicy Baked   
135                 Very Vanilla Soymilk, Single Serve   
136                                 Soymilk, Chocolate   
137                               Very Vanilla Soymilk   
138                       Shelf Stable Vanilla Soymilk   
139                                    Vanilla Pudding   
140                                 Keebler Soft Batch   
141                                        M&M cookies   
142                                     tortilla chips   
143                                 Sun chips original   
144                   Oven baked scoops tortilla chips   
145                            Pita Chips Simply Naked   
146  Sunchips harvest cheddar flavored multigrain s...   
147                                           Tostitos   
148                                        Apple juice   
149  Morningstar Farms Veggie Burgers Spicy Black B...   
150  Morningstar Farms Breakfast Sausage Patties 1....   
153                             Dry roasted pistachios   
154                       Planters Peanuts Salted Tube   
156                        Planters Cashew Salted Tube   
157                         Nuts & Chocolate Trail Mix   
158                            Quick 1-Minute Oats imp   
159               Quaker quick minute oats whole grain   
160  Old fashioned 100% whole grain oats, old fashi...   
161            Keebler Grahams Crackers Original .78oz   
162                       Blueberry whole grain muffin   
163                               Gyro bread pre-oiled   
164                               Gyro bread pre-oiled   
165                                    Parmesan cheese   
166                           Low Fat Milk (Grab & Go)   
167  Hershey s milk chocolate with almonds candy ba...   
168                                 Peanut Butter Cups   
169                 Kit Kat Duos Mint + Dark Chocolate   
170                                     Miniature Cups   
171                          Reese mini cups king size   
172                            breathsavers peppermint   
173                      Ultimate cheddar cheese sauce   
174                                Mild Cheddar Cheese   
175                  Welchs fruit snacks island fruits   
176                         Welch's Berries N Cherries   
177                           Mixed Fruit Fruit Snacks   
178                              Long Grain Brown Rice   
179           Old World Style Traditional Sauce(cajas)   
180     Special k protein bars chocolate peanut butter   
181                                  Protein meal bars   
182  Kellogg's Special K Protein Bars Double Chocol...   
183                                 Sour Cream & Onion   
184                                     Cheddar Cheese   
185                                     Pringles Ranch   
186                                     Pringles Ranch   
187      Kellogg's Eggo Waffles Homestyle 1.23oz 144ct   
188                                      Potato crisps   
189         Sweetened cereal with apple & cinnamon imp   
190                      Frosted flakes of corn cereal   

                           brands  \
128                      Sunshine   
129                      Cheez It   
130                      Cheez It   
131                      Cheez It   
132                      Cheez-It   
133                      Sunshine   
134                      Cheez It   
135                          Silk   
136  Silk,  Wwf Operating Company   
137  Silk,  Wwf Operating Company   
138                          Silk   
139                    Snack Pack   
140                   

In [81]:
merged.query("is_good_match")[["product_name", "brands", "categories", "ingredients_text", "Vendor", "Brand Name", "Product Type", "Food Product Category", "Product Name"]].iloc[120:180]

product_name  \
191                                    pop. tart bites   
192         Pop-tarts Bites Frosted Strawberry peg bag   
193          Pop tarts bites chocolately fudge peg bag   
195                          Nutrigrain Apple Cinnamon   
196                             Nutri Grain Strawberry   
197          Frosted hot fudge sundae toaster pastries   
198                          Breakfasttoaster pastries   
199                          Breakfasttoaster pastries   
200                                   Poptarts S'mores   
202                  Krave cereal with flavored center   
203                Pringles Sour Cream & Onion (small)   
204                            Pringles Cheddar Cheese   
205                       Grab and go sour cream onion   
206            Kellogg s breakfast cereal in a cup imp   
207                             Low Sodium Pinto Beans   
208                                        Black Beans   
209             Almondmilk Original Unsweetened 64 oz.   
210                                 Starburst original   
211                                    Snickers Almond   
212                                           Skittles   
213                         Skittles tropical standard   
214  Milk chocolate, peanuts, caramel, nougat compl...   
215                                   Twix Cookie Bars   
216                                   3 Musketeers S.S   
217                                     Milk Chocolate   
218                     peanut m&m's chocolate candies   
219                     peanut m&m's chocolate candies   
220                             Mm's chocolate candies   
221                                          Starburst   
222                                           milkyway   
223                                           Snickers   
224                                 milk chocolate m&m   
225                  Twix cookies and creme share size   
226                                     Ranch Dressing   
227                                           Dressing   
228                             creamy ceaser dressing   
229                            Golden italian dressing   
230                             Extra Heavy Mayonnaise   
231                            Kens Steak House Caesar   
232                          Pan asian sesame dressing   
233                                      Honey mustard   
234                                              Bites   
235                                      Fat Free Milk   
236                                            Lactaid   
237                                      Fat free milk   
238                                 Marinade and sauce   
239                               Worcestershire sauce   
240                      Original cayenne pepper sauce   
241                      Welch's, jelly, concord grape   
242                                   Haribo Goldbears   
243                              Haribo Goldbears Sour   
244                                  Frogs Gummi Candy   
245                                         Watermelon   
246                     Pizza rolls pizza rolls snacks   
247               Totinos pepperoni Pizza rolls 7.5 oz   
248     Big League Chew Bubble Gum Slammin’ Strawberry   
249                                          Oreo Mini   
250                                      Ritz crackers   
251                      Chips ahoy original shelf imp   
252  Nabisco lorna doone cookies-convenience pack 1...   

                                                brands  \
191                                                NaN   
192                                          Pop-Tarts   
193                                                NaN   
195                                          Kellogg's   
196                                          Kellogg's   
197                                          Kellogg's   
198                                                NaN   
199                                    

In [82]:
merged.query("is_good_match")[["product_name", "brands", "categories", "ingredients_text", "Vendor", "Brand Name", "Product Type", "Food Product Category", "Product Name"]].iloc[180:240]

product_name  \
253                                Easy Cheese Cheddar   
254                     Nabisco Chips Ahoy! Mini Chewy   
255                                   Golden Oreo mini   
256                                    Original Hommus   
257               Buttermist Food Spray & Pan Releaser   
258                                        Cooked Rice   
259                                       Sugar Packet   
260                 Premium Pure Cane Granulated Sugar   
261                 Premium pure cane granulated sugar   
262                                            Egg nog   
263                                General tso's sauce   
264                        Minor's, zesty orange sauce   
265                      Chicken Noodle Condensed Soup   
266        Campbellscondensed new england clam chowder   
267                                        Traditional   
268                               Smuckers Grape Jelly   
269                                   Strawberry jelly   
270                                  Jif Peanut Butter   
271                                  Granulated garlic   
272                   8th Continent, Soymilk, Original   
273                                       Toilet paper   
274                                   One Sweet Tomato   
275                Buttery toffee popcorn with peanuts   
276                               Popcorn with peanuts   
277                                          Pedialyte   
278                                        wiener buns   
279                   Kielbasa,Polska,Gm,5:1,6"",10#,Z   
280                              Top ramen beef flavor   
281        Top Ramen Chicken Flavour Ramen Noodle Soup   
282                         Cup Noodles Chicken Flavor   
283                                  Chow Mein Chicken   
284  CHOW MEIN TERIYAKI BEEF FLAVOR RESTAURANT STYL...   
286               Swiss Miss marshmallow hot cocoa mix   
287                                     Organic Garlic   
288                       Mike and Ike Original Fruits   
289                      Assorted fruit gumdrops candy   
290                     Organic Extra Virgin Olive Oil   
291                             100% Whole Wheat Bread   
292                      Natural long grain brown rice   
293                      Natural long grain brown rice   
295                       Original fruit gummies candy   
296                          Strawberry Sugar Free Gum   
297                                      Sugarfree Gum   
298                               Natural Rice Vinegar   
299          Gone dilly fresh whole kosher dill pickle   
300                 Mochiko Blue Star Sweet Rice Flour   
301                   Cilantro lime flavor concentrate   
302                   Cilantro lime flavor concentrate   
303            Fire roasted poblano flavor concentrate   
304                                        Turkey base   
305                                          Beef base   
306                            Low sodium chicken base   
307                                       Orange Juice   
308                                         Quick oats   
309                                     Tzatziki Sauce   
310                                          Hot Sauce   
311                               Act II butter lovers   
312                                            2.75 Oz   
313                               ACT II Butter Lovers   
314                                          Emergen-C   

                                     brands  \
253                                 Nabisco   
254                                 Nabisco   
255                                    Oreo   
256                                 Cedar's   
257                                     NaN   
258                          Golden Pheonix   
259                                  Domino   
260                                  Domino   
261                                  Domino   
262                              

In [83]:
merged.query("is_good_match")[["product_name", "brands", "categories", "ingredients_text", "Vendor", "Brand Name", "Product Type", "Food Product Category", "Product Name"]].iloc[240:300]

product_name  \
315                              Whole Grain Spaghetti   
316                                       Icy Lemonade   
317                             Light Multigrain Bread   
319                                      Salted Butter   
320                       Nerds gummy clusters peg bag   
322                                        Big burrito   
323                                 Tina’s Big Burrito   
324                                        Big Burrito   
325                                       Hawaiian Sun   
326                                Pass o guava Nectar   
327                                    Iced Tea, Lemon   
328                                            Lilikoi   
329                            Distilled White Vinegar   
330                       Apple Cider Flavored Vinegar   
331                          Chunk light tuna in water   
332                         Lunch-To-Go Tuna Salad Kit   
333               Snake club sweet and salty trail mix   
334                                    Fancy trail mix   
335                                         Whole milk   
336                                         Whole Milk   
337                                         Whole milk   
338                                        Lowfat milk   
339                         Cream-O-Land, Fatfree Milk   
340                                2% Reduced Fat Milk   
341                                     Chocolate Milk   
342                      Cream-O-Land, Milk, Chocolate   
343                                             Yogurt   
344                             Sea Salt Medium Grains   
345                     Soup & Chowder Oyster Crackers   
347                  Sour Cream & Cheddar Potato Chips   
348                                       Potato chips   
349                                    Shelled Edamame   
350           Tgi fridays Cheddar & bacon potato skins   
351                       Sunflower seed creamy spread   
352                                   SunButter Creamy   
353                       Tomatillos - Whole Tomatillo   
354  Fully cooked and browned sausage links skinles...   
355                  Beyond Burger Plant Based Patties   
356                                              Slush   
357                         Chocolate Cupcakes Hostess   
358  Morningstar Farms Veggie Burgers Garden Veggie...   
359                             Bacn,Fc,Sm,Reg,2/150Ct   
360                             Bacn,Fc,Sm,Reg,2/150Ct   
361                 Savor Unseasoned Rice Wine Vinegar   
362                                 citavo grove juice   
363    Gardenburger Veggie Burgers Original 3.4oz 48ct   

                                         brands  \
315                                     Barilla   
316                                         NaN   
317                            Nature's Harvest   
319                                   Grassland   
320                                       Nerds   
322                                         NaN   
323                                         NaN   
324                                         NaN   
325                 Hawaiian Sun Products  Inc.   
326                 Hawaiian Sun Products  Inc.   
327  Hawaiian Sun,  Hawaiian Sun Products  Inc.   
328                                Hawaiian Sun   
329                             Supreme Oil Co.   
330                                         NaN   
331                                         NaN   
332                                    StarKist   
333                                   Snak Club   
334                                   Snak Club   
335                          Cream-O-Land Dairy   
336                          Cream-O-Land Dairy   
337                                         NaN   
338                                         NaN   
339                          Cream-O-Land Dairy   
340                          Cream-O-Land Dairy   
341                                Cream

In [97]:
merged.query("is_good_match").groupby(["nova_group", "Level of Processing"], dropna=False, as_index=False).size()

nova_group  Level of Processing  size
0          1.0                  1.0    20
1          1.0                  2.0     1
2          1.0                  4.0     1
3          2.0                  2.0     4
4          3.0                  1.0     2
5          3.0                  2.0     2
6          3.0                  3.0     5
7          3.0                  4.0     3
8          4.0                  1.0     1
9          4.0                  2.0     1
10         4.0                  3.0     3
11         4.0                  4.0   151
12         4.0                  NaN     1
13         NaN                  1.0    25
14         NaN                  2.0     7
15         NaN                  3.0     2
16         NaN                  4.0    56
17         NaN                  NaN     1

The confusion matrix:

|          | OFF = 1 |   2 |   3 |   4 |   ? |
|--:|--:|--:|--:|--:|--:|
| **CGFP = 1** |  20 |   0 |   2 |   1 |  25 |
|        **2** |   1 |   4 |   2 |   1 |   7 |
|        **3** |   0 |   0 |   5 |   3 |   2 |
|        **4** |   1 |   0 |   3 | 151 |  56 |
|        **?** |   0 |   0 |   0 |   1 |   1 |

Accuracy (among non-NaN, the upper-left 4×4 square) = 180/194 = 92.8%.

Agreement is pretty good, though NOVA categories 2 and 3 are very unrepresented. Open Food Facts is much more likely to leave a NOVA group unspecified, and biased toward NOVA = 0 when it does so (taking the CGFP assignments as truth).